pkg/client/typed/dynamic/client.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package dynamic provides a client interface to arbitrary Kubernetes
+// APIs that exposes common high level operations and exposes common
+// metadata.
+package dynamic
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+// Client is a Kubernetes client that allows you to access metadata
+// and manipulate metadata of a Kubernetes API group.
+type Client struct {
+	cl *client.RESTClient
+}
+
+// NewClient returns a new client based on the passed in config. The
+// codec is ignored, as the dynamic client uses it's own codec.
+func NewClient(conf *client.Config) (*Client, error) {
+	// avoid changing the original config
+	confCopy := *conf
+	conf = &confCopy
+
+	conf.Codec = &JSONCodec{}
+
+	if conf.APIPath == "" {
+		conf.APIPath = "/api"
+	}
+
+	if len(conf.UserAgent) == 0 {
+		conf.UserAgent = client.DefaultKubernetesUserAgent()
+	}
+
+	if conf.QPS == 0.0 {
+		conf.QPS = 5.0
+	}
+	if conf.Burst == 0 {
+		conf.Burst = 10
+	}
+
+	cl, err := client.RESTClientFor(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{cl: cl}, nil
+}
+
+// Resource returns an API interface to the specified resource for
+// this client's group and version. If resource is not a namespaced
+// resource, then namespace is ignored.
+func (c *Client) Resource(resource *unversioned.APIResource, namespace string) *ResourceClient {
+	return &ResourceClient{
+		cl:       c.cl,
+		resource: resource,
+		ns:       namespace,
+	}
+}
+
+// ResourceClient is an API interface to a specific resource under a
+// dynamic client.
+type ResourceClient struct {
+	cl       *client.RESTClient
+	resource *unversioned.APIResource
+	ns       string
+}
+
+// namespace applies a namespace to the request if the configured
+// resource is a namespaced resource. Otherwise, it just returns the
+// passed in request.
+func (rc *ResourceClient) namespace(req *client.Request) *client.Request {
+	if rc.resource.Namespaced {
+		return req.Namespace(rc.ns)
+	}
+	return req
+}
+
+// List returns a list of objects for this resource.
+func (rc *ResourceClient) List(opts api.ListOptions) (*List, error) {
+	result := new(List)
+	err := rc.namespace(rc.cl.Get()).
+		Resource(rc.resource.Name).
+		VersionedParams(&opts, api.Scheme). // What to do about VersionedParams converter?
```

Comment:
There is no generic codec yet - just giving it that name for convenience
here.  dynamic.ParameterCodec

On Wed, Feb 3, 2016 at 1:33 PM, krousey <notifications@github.com> wrote:

> In pkg/client/typed/dynamic/client.go
> <https://github.com/kubernetes/kubernetes/pull/20351#discussion_r51765129>
> :
>
> > +// namespace applies a namespace to the request if the configured
> > +// resource is a namespaced resource. Otherwise, it just returns the
> > +// passed in request.
> > +func (rc *ResourceClient) namespace(req *client.Request) *client.Request {
> > +	if rc.resource.Namespaced {
> > +		return req.Namespace(rc.ns)
> > +	}
> > +	return req
> > +}
> > +
> > +// List returns a list of objects for this resource.
> > +func (rc *ResourceClient) List(opts api.ListOptions) (*List, error) {
> > +	result := new(List)
> > +	err := rc.namespace(rc.cl.Get()).
> > +		Resource(rc.resource.Name).
> > +		VersionedParams(&opts, api.Scheme). // What to do about VersionedParams converter?
>
> I'm having trouble finding the ParameterCodec you mention. The only
> EncodeParameters implementation is on runtime.parameterCodec and the only
> call to runtime.NewParameterCodec is in pkg/api/register.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351/files#r51765129>.
>